### PR TITLE
WIP: Add a first graph topology serialization using ONNX format

### LINKED ISSUE #296: Model Serialization
We need to be able to serialize models in OpenMined. This needs to be done to send models over the wire and potentially to IPFS. Probably a few different ways to achieve this. 

Native C#: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/serialization/

Pros: Built in
Cons: Is this supportable across languages?

Protobufs: https://github.com/google/protobuf

Pros: Portable
Cons: Third party dependency

ONNX: https://github.com/onnx/onnx (this one is probably not an option for now, just putting it here for discussion sake)

Pros: Will be handy to have this one day so we can import and export models to other frameworks
Cons: A lot of heavy lifting left to do to support this

I don't know much about any of these options so feel free to add/correct anything! 😄

### PR DESCRIPTION
**Pull request by morgangiraud**
# Description

Hi everyone,

I've implemented a very raw version of graph serialisation using the JSON format.
You can look at this [notebook](https://github.com/morgangiraud/OpenMined/blob/feature/serialization/notebooks/keras/Keras%20serialization.ipynb) to see the result of the PySyft & OpenMined work so far.
More information can be find at this issue's thread: #296

Fixes #296 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

You can find a pseudo-test in this [notebook](notebooks/keras/Keras%20serialization.ipynb)
But this is currently a work in progress, no real tests have been made so far.

**Pseudo-Test Configuration**:
* CPU: Ok

# Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by bendecoste** [comment]
isn't newtonsoft already in the project?

**Comment by morgangiraud** [comment]
Yes, it was. But couldn't make it working in OSX.

Someone has made a [Unity version of Newtonsoft.JSON](https://www.assetstore.unity3d.com/en/#!/content/11347) which worked like a charm for me. 
I'm completely new to C# and Unity, so feel free to correct me if I'm mistaken and I though it was fine to replace the plugin with unity version.

Of course, I just hope, this will be fine for everybody else too 😅

File: UnityProject/Assets/OpenMined/Network/Controllers/SyftController.cs [hunk 1]
```diff
@@ -492,6 +495,27 @@ public string processMessage (string json_message, MonoBehaviour owner)
...
+
+							var input = File.OpenRead(filename);
+							ModelProto modelProto = ModelProto.Parser.ParseFrom(input);
+						}
+						else if (msgObj.functionCall == "to_proto")
+						{
+
+							ModelProto model =  this.ToProto(msgObj.tensorIndexParams);
+							string filename = msgObj.tensorIndexParams[2];
+							// using (var output = new CodedOutputStream(File.Create(filename)))
```
Comment (sagivo) [inline UnityProject/Assets/OpenMined/Network/Controllers/SyftController.cs]: is it needed? if not, we can leave it out

File: UnityProject/Assets/OpenMined/Network/Controllers/SyftController.cs [hunk 2]
```diff
@@ -520,6 +544,31 @@ public string processMessage (string json_message, MonoBehaviour owner)
 			return "Unity Error: SyftController.processMessage: Command not found:" + msgObj.objectType + ":" + msgObj.functionCall;
 		}
 
+		private ModelProto ToProto (string[] Params)
+		{
+			int model_id = int.Parse(Params[0]);
```
Comment (sagivo) [inline UnityProject/Assets/OpenMined/Network/Controllers/SyftController.cs]: important to keep consistency between camelCase and snake_case in c# variable names. usually, it's cammelCase.

File: UnityProject/Assets/OpenMined/Network/Controllers/SyftController.cs [hunk 3]
```diff
@@ -520,6 +544,31 @@ public string processMessage (string json_message, MonoBehaviour owner)
 			return "Unity Error: SyftController.processMessage: Command not found:" + msgObj.objectType + ":" + msgObj.functionCall;
 		}
 
+		private ModelProto ToProto (string[] Params)
```
Comment (sagivo) [inline UnityProject/Assets/OpenMined/Network/Controllers/SyftController.cs]: maybe a more descriptive name? also, we should follow the c# guidelines for naming (function params are should not be capitalized).

File: UnityProject/Assets/OpenMined/Protobuf/onnx-operators.proto3 [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  // numbers no greater than opset_version.
+  int64 opset_version = 5;
+
+  // A human-readable documentation for this set of operators. Markdown is allowed.
+  string doc_string = 6;
+
+  // The operators specified by this operator set.
+  // The (name, version) MUST be unique across all OperatorProtos in operator
+  repeated OperatorProto operator = 8;
+}
```
Comment (sagivo) [inline UnityProject/Assets/OpenMined/Protobuf/onnx-operators.proto3]: nl

File: UnityProject/Assets/OpenMined/Protobuf/OnnxTools.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                Node = { node },
+                Initializer = { inits },
+                Input = { infos },
+            };
+
+            return g;
+        }
+    }
+}
```
Comment (sagivo) [inline UnityProject/Assets/OpenMined/Protobuf/OnnxTools.cs]: nl

**Pull request merged**